CMake: Improve stl checks

MaartenBent opened the pull request:
see https://trac.wxwidgets.org/ticket/18718

MaartenBent committed:
CMake: Set wxBUILD_CXX_STANDARD globally instead of only on targets

This way the tests in setup.cmake will use the correct c++ version.

MaartenBent committed:
CMake: Add variable to indicate if c++11 or later is enabled

Use it in macOS checks.

MaartenBent committed:
CMake: Only test stl for c++ versions before 11

vadz commented on build/cmake/init.cmake:
```diff
@@ -8,6 +8,15 @@
+if(DEFINED wxBUILD_CXX_STANDARD AND NOT wxBUILD_CXX_STANDARD STREQUAL COMPILER_DEFAULT)
+    set(CMAKE_CXX_STANDARD ${wxBUILD_CXX_STANDARD})
+endif()
+if(CMAKE_CXX_STANDARD EQUAL 11 OR CMAKE_CXX_STANDARD EQUAL 14 OR CMAKE_CXX_STANDARD EQUAL 17)
```
Is there really no way to say `>= 11` in CMakese?

MaartenBent replied:
Yes, I tried. But then 98 is also >11.

vadz replied:
Oh. But it seems that 20 is also [already supported](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html#prop_tgt:CXX_STANDARD), so maybe we should test for `!= 98` then?

MaartenBent replied:
Yes, that is better.

I see MSVC now also supports setting the c++ version, I'll enable it for this compiler as well.

The build however fails on `std::bind2nd` because it still reports `__cplusplus 199711L` (`/Zc:__cplusplus` is not defined by CMake https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). I'll add `wxCHECK_VISUALC_VERSION(10)` to the check.

vadz replied:
> The build however fails on `std::bind2nd` because it still reports `__cplusplus 199711L` 

Ugh, yes, of course, sorry for missing this.

> I'll add `wxCHECK_VISUALC_VERSION(10)` to the check.

I'm pretty sure we can use `std::bind2nd` up until MSVS 2019 if we really want to, but I don't think MSVC 10 supported range-for loops, I believe this was added in 11.

MaartenBent replied:
> I'm pretty sure we can use std::bind2nd up until MSVS 2019 if we really want to, but I don't think MSVC 10 supported range-for loops, I believe this was added in 11.

Thanks. I only checked the `auto` keyword, not the range. In adition, it should also have support for lambda functions (`wxArrayString::Sort` and `wxSortedArrayString::Index`). Since the `/std:c++` switch is only available since VS2015u2, I'll be on the save side and use `wxCHECK_VISUALC_VERSION(14)`.

While looking at the `__cplusplus` usages, I also noticed that the `wxMEMBER_DELETE = delete` is not used for MSVC. It is also available since VS2015. Same for `wxUniCharRef(const wxUniCharRef&) = default`.

I'll test some more and update the PR later.

vadz reviewed (approved):
Thanks, will merge once the CI builds pass.

MaartenBent committed:
CMake: Add c++20 option, enable setting c++ version for MSVC

MSVC supports setting the CMAKE_CXX_STANDARD since VS2015 update 2.

MaartenBent committed:
Fix wxArrayString build error when using c++17 with MSVC

MSVC does not set the correct __cplusplus version, so check MSVC compiler
version as well.

MaartenBent committed:
Fix building test.cpp with c++17

uncaught_exception() has been replaced with uncaught_exceptions().

MaartenBent committed:
Enable wxMEMBER_DELETE for MSVC

It is available since Visual Studio 2015.

MaartenBent committed:
Define wxMEMBER_DEFAULT similarly to the existing wxOVERRIDE

Also define wxHAS_MEMBER_DEFAULT allowing to check if '= default' is supported
(compilers that define c++11 and VS2015 and later).

MaartenBent committed:
Fix previous local declaration warning in wxRegKey sample

MaartenBent commented:
Can you think of a better solution for https://github.com/wxWidgets/wxWidgets/pull/1782/commits/93e0e5ff0be2ac53dd8dbfb53316002b73ef6936? 

It uses copy in `TempStringRef::Make`, while the class was marked as `wxDECLARE_NO_COPY_CLASS`:
```c++
const TempStringRef TempStringRef::Make(const wxString &str)
{
    return TempStringRef(str);
}
```

vadz commented:
I don't really like allowing copy-constructing it because this is pretty dangerous, e.g. it would allow code like this:
```cpp
rt::TempStringRef t(rt::TempStringRef::Make("BOOM"));
```
to compile and, presumably, crash when trying to use the string in any way.

The simplest solution I see is to just remove `Make()` entirely, I don't see what benefit does it have compared to simply using the ctor. @TcT2k do you see anything wrong with it?

PBfordev commented on build/cmake/options.cmake:
```diff
@@ -36,15 +36,17 @@ endif()
-    # Other compilers support setting the C++ standard, present it an option to the user
+endif()
+
+if(NOT MSVC OR NOT MSVC_VERSION GREATER 1900)
```
"**OR NOT** MSVC_VERSION GREATER 1900" Does the "NOT" really belong there?

Additionally, MSVC does not support all C++ standards listed below (only 14, 17, and latest) but I guess this applies to other compilers as well...

MaartenBent replied:
That's indeed wrong. I wanted to change it to `NOT MSVC_VERSION LESS 1900` but I guess I'll just keep `MSVC_VERSION GREATER 1800` for simplicity (`GREATER_EQUAL` is not available in older CMake versions).

MaartenBent committed:
Fix using deleted function in TempStringRef

Use the constructor directly instead of via a make function.

MaartenBent committed:
CMake: install toucan.png with widgets sample

It is used on the StaticBitmap page.

vadz commented on include/wx/animate.h:
```diff
@@ -38,9 +38,9 @@ class WXDLLIMPEXP_CORE wxAnimation : public wxObject
-#if __cplusplus >= 201103
-    wxAnimation(const wxAnimation&) = default;
-    wxAnimation& operator=(const wxAnimation&) = default;
+#ifdef wxMEMBER_DEFAULT
```
I'd like to write this as
```suggestion
#ifdef wxHAS_MEMBER_DEFAULT
```
for consistency with the other similar macros. Also, by a coincidence, I've just added `wxHAS_NOEXCEPT` and `wxNOEXCEPT` in #1788, and it would really make sense to handle them in the same way.

MaartenBent replied:
Maybe I can remove `wxHAS_MEMBER_DEFAULT` entirely? Guarding the function declaration does not generate build warnings or errors, only an Intellisense warning about 'function definition not found'.

vadz replied:
This would definitely generate link errors if either of the special methods is actually used with pre-C++11 compiler.

This is the problem, `wxMEMBER_DEFAULT` and `wxMEMBER_DELETE` are not symmetrical: the latter can be used unconditionally, but the former really can't be.

vadz commented on include/wx/unichar.h:
```diff
@@ -267,8 +267,8 @@ class WXDLLIMPEXP_BASE wxUniCharRef
 
-#if __cplusplus >= 201103
-    wxUniCharRef(const wxUniCharRef&) = default;
+#ifdef wx_MEMBER_DEFAULT
```
This looks like a typo?
```suggestion
#ifdef wxMEMBER_DEFAULT
```

vadz commented on include/wx/unichar.h:
```diff
@@ -267,8 +267,8 @@ class WXDLLIMPEXP_BASE wxUniCharRef
-#if __cplusplus >= 201103
-    wxUniCharRef(const wxUniCharRef&) = default;
+#ifdef wx_MEMBER_DEFAULT
+    wxUniCharRef(const wxUniCharRef&) wx_MEMBER_DEFAULT;
```
```suggestion
    wxUniCharRef(const wxUniCharRef&) wxMEMBER_DEFAULT;
```
